### ISSUE TITLE
net.razorvine.pickle.InvalidOpcodeException: invalid pickle opcode: 120

### ISSUE THREAD

**Issue by vdmitriyev** [jpmml/sklearn2pmml#8]
I'm receiving following error after installing everything properly and running given in README.md example. 

```
Jul 20, 2016 4:46:11 PM org.jpmml.sklearn.Main run
INFO: Parsing DataFrameMapper PKL..
Jul 20, 2016 4:46:11 PM org.jpmml.sklearn.Main run
SEVERE: Failed to parse DataFrameMapper PKL
net.razorvine.pickle.InvalidOpcodeException: invalid pickle opcode: 120
        at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:330)
        at net.razorvine.pickle.Unpickler.load(Unpickler.java:99)
        at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:230)
        at org.jpmml.sklearn.Main.run(Main.java:126)
        at org.jpmml.sklearn.Main.main(Main.java:107)

Exception in thread "main" net.razorvine.pickle.InvalidOpcodeException: invalid pickle opcode: 120
        at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:330)
        at net.razorvine.pickle.Unpickler.load(Unpickler.java:99)
        at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:230)
        at org.jpmml.sklearn.Main.run(Main.java:126)
        at org.jpmml.sklearn.Main.main(Main.java:107)
Traceback (most recent call last):
```

Thank you for your effort.

**Comment by vruusmann** [comment]
I'll try to collect as much information as possible, but I'm afraid that this issue needs to be dispatched to https://github.com/irmen/Pyrolite later on.

What is your Python version, package versions?
```
import sklearn, pandas, sklearn_pandas, joblib, numpy

print(sklearn.__version__)
print(pandas.__version__)
print(sklearn_pandas.__version__)
print(joblib.__version__)
print(numpy.__version__)
```

The `sklearn2pmml` package has been tested with Python 2.7 and 3.4, using the latest package versions.

Can you attach the problematic `DataFrameMapper` PKL file here?

**Comment by vdmitriyev** [comment]
I used Anaconda 4.1.1 (64-bit). Here are versions I have:
- Python 3.5.2 
- sklearn - 0.17.1
- pandas - 0.18.1
- sklearn_pandas - 1.1.0
- joblib - 0.10.0
- numpy - 1.11.1

Unfortunately I cannot attch DataFrameMapper. I already removed everything and trying to reuse other stack.

**Comment by vdmitriyev** [comment]
This time I tried on Ubuntu machine. When parameter 'with_repr = False' launcher command generates (I mean ```sklearn2pmml``` method) a command line that is now valid. In case I separately execute command line I'm getting ```-bash: syntax error near unexpected token `('``` error. Here is command I tried.

```
$ java -cp /home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/slf4j-api-
1.7.21.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jaxb-runtime-2.2.11.jar:/home/vagran
t/.local/lib/python2.7/site-packages/sklearn2pmml/resources/slf4j-jdk14-1.7.21.jar:/home/vagrant/.local/lib/python2.7/s
ite-packages/sklearn2pmml/resources/jpmml-sklearn-1.0-SNAPSHOT.jar:/home/vagrant/.local/lib/python2.7/site-packages/skl
earn2pmml/resources/pmml-agent-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-
schema-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pyrolite-4.12.jar:/home/vagra
nt/.local/lib/python2.7/site-packages/sklearn2pmml/resources/guava-19.0.jar:/home/vagrant/.local/lib/python2.7/site-pac
kages/sklearn2pmml/resources/pmml-model-metro-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/
resources/jaxb-core-2.2.11.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jcommander-1.48.
jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/istack-commons-runtime-2.21.jar:/home/vagra
nt/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jpmml-converter-1.0.7.jar:/home/vagrant/.local/lib/python2
.7/site-packages/sklearn2pmml/resources/serpent-1.12.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/
resources/jpmml-xgboost-1.0.5.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-model-1.
2.16.jar org.jpmml.sklearn.Main --pkl-estimator-input /tmp/tmpJojwrR.pkl --repr-estimator LogisticRegressionCV(Cs=10, c
lass_weight=None, cv=None, dual=False,           fit_intercept=True, intercept_scaling=1.0, max_iter=100,           mul
ti_class='ovr', n_jobs=1, penalty='l2', random_state=None,           refit=True, scoring=None, solver='lbfgs', tol=0.00
01, verbose=0) --pkl-mapper-input /tmp/tmpAAxd0m.pkl --repr-mapper DataFrameMapper(features=[(['Sepal.Length', 'Sepal.W
idth', 'Petal.Length', 'Petal.Width'], TransformerPipeline(steps=[('continuousdomain', ContinuousDomain(invalid_value_t
reatment='return_invalid')), ('pca', PCA(copy=True, n_components=3, whiten=False))])), ('Species', None)],        spars
e=False) --pmml-output LogisticRegressionIris.pmml
```

I also think that in "verbose" mode you need to use `print(' '.join(cmd))` instead of `print(cmd)`. I also think that you need an additional *debug* option that won't delete *pkl* file and will show command line to execute it manually.

**Comment by vdmitriyev** [comment]
Here is error exception I'm getting now from Java:

```
Exception in thread "main" java.lang.UnsupportedClassVersionError: org/jpmml/sklearn/Main : Unsupported major.minor version 51.0
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:643)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at java.net.URLClassLoader.defineClass(URLClassLoader.java:277)
        at java.net.URLClassLoader.access$000(URLClassLoader.java:73)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:212)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:205)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:323)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:296)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:268)
        at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:406)
```

Here is command that starts everything (generated by ```sklearn2pmml``` method):
```
java -cp /home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/slf4j-api-1.7.21.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jaxb-runtime-2.2.11.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/slf4j-jdk14-1.7.21.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jpmml-sklearn-1.0-SNAPSHOT.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-agent-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-schema-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pyrolite-4.12.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/guava-19.0.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-model-metro-1.2.16.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jaxb-core-2.2.11.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jcommander-1.48.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/istack-commons-runtime-2.21.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jpmml-converter-1.0.7.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/serpent-1.12.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/jpmml-xgboost-1.0.5.jar:/home/vagrant/.local/lib/python2.7/site-packages/sklearn2pmml/resources/pmml-model-1.2.16.jar org.jpmml.sklearn.Main --pkl-estimator-input /tmp/tmpuuG9iL.pkl --pkl-mapper-input /tmp/tmp9z6pYr.pkl --pmml-output LogisticRegressionIris.pmml
```

DataFrameMapper:
```
DataFrameMapper(features=[(['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width'], TransformerPipeline(steps=[('continuousdomain', ContinuousDomain(invalid_value_treatment='return_invalid')), ('pca', PCA(copy=True, n_components=3, whiten=False))])), ('Species', None)],        sparse=False)
```

I also attached PKL files.

[sklearn2pmml-pkl-files.zip](https://github.com/jpmml/sklearn2pmml/files/375827/sklearn2pmml-pkl-files.zip)

**Comment by vruusmann** [comment]
The `Unsupported major.minor version 51.0` exception means that your Ubuntu machine contains Java 1.6.0, which violates JPMML-SkLearn requirement of Java 1.7.0+.

**Comment by vruusmann** [comment]
Just double-checked the `LogisticRegressionIris` example with Python 2.7.11 and Python 3.4.3 (on Fedora release 23), and everything works as intended.

For comparison, here are my package versions:
```
sklearn2pmml - 0.9.4
sklearn - 0.17.1 (Python 2.7) and 0.18.dev0 (Python 3.4)
pandas - 0.18.0
sklearn_pandas - 1.1.0
joblib - 0.9.4
numpy - 1.11.0
```

The only major difference between my and your environment is the `joblib` package version. Yours is newer, so perhaps it does the PKL serialization of SkLearn estimators differently. I will upgrade to `joblib` 0.10.0 later today and see if it breaks my workflow or not.

In the meantime, you could downgrade to `joblib` 0.9.4 and see if that helps.

**Comment by vdmitriyev** [comment]
Many thanks for your prompt support. I totally overlooked java version and installed default one. Now I'm working to install a newer (proper) version of Java and trying everything one  more time.

**Comment by vruusmann** [comment]
Thanks for attaching sample PKL files - I get the same `invalid pickle opcode: 120` exception when I try to convert them using JPMML-SkLearn command-line application, so that's a good starting point for debugging work.

**Comment by vdmitriyev** [comment]
You are welcome. I just update Java to proper version and I'm still getting the same ```invalid pickle opcode: 120``` exception.

**Comment by vruusmann** [comment]
I can confirm that upgrading the `joblib` package from 0.9.4 to 0.10.0 causes the workflow to break with this `invalid pickle opcode: 120` exception.

**Comment by vdmitriyev** [comment]
Thank you a lot! It's finally worked for me! I'm attaching "requirements.txt" I used.

[requirements.txt](https://github.com/jpmml/sklearn2pmml/files/376039/requirements.txt)

**Comment by vruusmann** [comment]
I'm reopening this issue, because I definitely want to support all `joblib` package versions. I've done some research, and it looks like 0.10.0 is adding an extra layer of compression somewhere. The opcode 120 is a flag that informs consumers about it.

Hope to have this issue solved later this evening.

**Comment by vdmitriyev** [comment]
@vruusmann You are completely right in reopening this ticket.

**Comment by vruusmann** [comment]
I've got the fix ready in my computer. The release will be delayed by a day or two, because I want to implement a deserialization test matrix (Python 2.7/3.4 vs. Pickle/Joblib vs. Joblib 0.9/0.10.0 - What a mess!) to increase the robustness of it all.

**Comment by vdmitriyev** [comment]
@vruusmann Sounds great! I will shortly send you a pull request with debug options if you don't mind.

**Comment by denmase** [comment]
Hi @vruusmann, 

I'm sorry for bumping this old and closed issue up, but strangely I got the same error now, which haven't been occurred to me before. To make sure, I ran the LightGBM example from your blog entry. It resulted the same error.

Here is my environment:
`python-3.10.5.amd64

sklearn2pmml         0.86.3
scikit-learn         1.1.2
pandas               1.5.0
sklearn-pandas       2.2.0
joblib               1.2.0
numpy                1.23.3

java version "1.8.0_141"
Java(TM) SE Runtime Environment (build 1.8.0_141-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.141-b15, mixed mode)`.

Is there anything that I can do to investigate further? Thank you
